### PR TITLE
Default shape and context attributes for operators

### PR DESCRIPTION
**Pull request by reminisce**
This PR resolves the issues reported by @szha as the following:

1. Imperative random operators such as `mx.nd.random.uniform()` results in segmentation fault when used without user input shapes. That's because the shape attribute defaults to an empty shape if it's not provided by users. But in Numpy, such statement `np.random.uniform()` can run through and prints a single value. In order to keep consistent with Numpy's behavior, we need to default `shape=(1,)` if it is not provided by users for random operators through either `shape` or `out` attribute imperatively. For symbolic random operators, default shape is still an empty one since it might need to be inferred from other symbolic variables.

2. For the imperative operators with `ctx` as one of its attributes, it should default to the `context.current_context` instead of `mx.cpu(0)` if users did not specify through either `ctx` or `out` attribute, so that the following use case can be supported:
```python
with mx.Context(mx.gpu(0)):
    output = mx.nd.random.uniform()
    assert output.context == mx.current_context()  # current ctx is gpu(0)
```

@piiswrong @szha @eric-haibin-lin

### CODE REVIEW

File: python/mxnet/ndarray/op.py [hunk 1]
```diff
@@ -83,16 +91,29 @@ def _make_ndarray_function(handle, name):
...
     ndarg_names = []
     kwarg_names = []
+    is_random_op = _is_random_op(func_name)
     for i in range(narg):
         name, atype = arg_names[i], arg_types[i]
         if name == 'dtype':
             dtype_name = name
-            signature.append('%s=_Null'%name)
+            signature.append('%s=_Null' % name)
+        elif name == 'shape' and is_random_op:
```
Comment (piiswrong) [inline python/mxnet/ndarray/op.py]: Just change the default value to TShape(1) in the backend

**Comment by zhreshold** [comment]
Just curious how can random_op infer shape from other variables? 
A default shape=(1,) should be fine otherwise.

File: python/mxnet/ndarray/op.py (hunk 1 above)
Comment (piiswrong) [reply python/mxnet/ndarray/op.py]: ok I see the problem here. still don't quite like the is_random_op solution though.

**Comment by reminisce** [comment]
@zhreshold I tried setting `shape=(1,)` in backend, but many symbolic test cases failed, because the shape defaults to `(1,)` while the output's shape was inferred from the op downstream.

File: python/mxnet/ndarray/op.py (hunk 1 above)
Comment (mli) [reply python/mxnet/ndarray/op.py]: agree it is a little bit hacky, but let merge a runnable version first

Comment (reminisce) [reply python/mxnet/ndarray/op.py]: Ok. I will submit another PR making the changes in the way as @piiswrong suggested after my another work is finished.

**Pull request closed without merging**